### PR TITLE
Suggestion Skillframework v2.1

### PR DESCRIPTION
**Pull request by forslund**
## Description
Currently the skill is is loaded in several steps where the skill-loader code calls the startup sequence and doesn't provide the working skill until after calling `bind()`, `load_data_files()`, etc. All which really should be part of the skill `__init__()`. Skill creator are left with doing their init code in `initialize` a unique method just used by skills and not common across python.

Current skill system also requires some further boiler plate in the form of the `create_skill()` function instantiating the skill and returning it to the SkillLoader. 

This is a suggestion to allow the skill creators to do all their custom initialization in `__init__()` like a regular python class and remove the extra `create_skill()`boilerplate.

 To be backwards compatible this keeps two skill loading functions in the SkillLoader `_create_v2_skill_instance()` and `_create_v3_skill_instance()`, the v2 version will be used if create_skill is found in the skill module, if no such function is found and there is a MycroftSkill based object the loader will consider this a v3 skill module.

Requirements of the v3 skill:

For this to work the skill creator will need to setup their skill `__init__()` (if used) to forward the keyword arguments to the base class.

Example:
```python
from mycroft import MycroftSkill, skill_handler


class MySkill(MycroftSkill):
    def __init__(**kwarg):
        super().__init__(name='My Cool Skill', **kwarg)  # Name is of course optional
        [ Further init code ]

    @skill_handler('Cool.intent')
    def cool_handler(self, _):
        self.speak('My skill is so cool!')
```

This requirement to pass along kwarg could also be avoided with a custom `__new__()` but I think it's better to have an explicit pass-along so skill creators are aware that Mycroft is doing stuff.

Since we now get a class from the skill and not a loaded object the class can provide compatibility information. In this proposal the skill creator can have several skill classes in their module and use the class method is_compatible to select which of the provided classes to use.

In summary this would reduce some of the boiler plate code needed and make the skill loading more pythonic.

## Contributor license agreement signed?
CLA [ yes ]

### CODE REVIEW

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2855))

**Comment by krisgesling** [comment]
Seems like it would be a good shift, however if we're making such a significant change to how Skills are written, I think we should consider if there are other architectural aspects that we want to modify at the same time. If we can avoid supporting multiple versions of the Skills framework (could easily end up with "v2, v3, v4, v5", ... ) then it will be much easier for us in the long run.

One of the key ones in my mind is decoupling Skills from core. How might we provide an architecture that allows Skills to be completely independent from the rest of the system, communicating with core only via the message bus?

The primary benefit being that we can then isolate Skills more easily. However I think this would also make testing of an individual Skill easier.

**Comment by JarbasAl** [comment]
this doesnt really seem like `v3`, but rather like a general cleanup and making the `create_skill` method optional, it doesnt deprecate anything and is fully backwards compatible, i would go with `v2.1` nomenclature 

i want to point out that the `create_skill` method is potentially very useful if some skill needs to perform some action before the loading stage for some reason, so i think it's good to keep support for it

@krisgesling comment is relevant, but feels like a separate discussion and follow up PR since this isn't really a significant change (for skill devs)

**Comment by forslund** [comment]
@JarbasAI you're correct this is more of a cleanup. I'll drop the 3 and call it `2_1` in the code next time I update. Regarding the the comments about create_skill couldn't you in most cases just run your "before skill loading" the `super().__init__()` call?

You could also implement the __new__() if there are truly things that require that the skill object isn't instantiated at all or have a section in the skill that is run on import?

@krisgesling I do think we'd always need to support 2 versions in some way, The current and the deprecated, when a new current is added the deprecated is removed and the old current is made deprecated.

Regarding the skill decoupling, Do we have a list of these hard mycroft system connections? I know of the fallback system system (issue #2806). Are you thinking of the mycroft utils folder and such?

File: mycroft/skills/mycroft_skill/mycroft_skill.py [hunk 1]
```diff
@@ -63,6 +63,10 @@
 )
 
 
+def skill_id_from_path(skill_directory):
+    return basename(skill_directory)
+
+
```
Comment (chrisveilleux) [inline mycroft/skills/mycroft_skill/mycroft_skill.py:69]: rather than making this a separate function, why not make skill_id a property?

File: mycroft/skills/mycroft_skill/mycroft_skill.py [hunk 2]
```diff
@@ -165,6 +168,32 @@ def __init__(self, name=None, bus=None, use_settings=True):
         # Skill Public API
         self.public_api = {}
 
+        if bus:
```
Comment (chrisveilleux) [inline mycroft/skills/mycroft_skill/mycroft_skill.py:171]: What happens if the bus is not passed in?

File: mycroft/skills/mycroft_skill/mycroft_skill.py [hunk 3]
```diff
@@ -165,6 +168,32 @@ def __init__(self, name=None, bus=None, use_settings=True):
...
+            self.load_data_files()
+            # Set up intent handlers
+            self._register_decorated()
+            self.register_resting_screen()
+            self.initialize()
+        except Exception as e:
+            # If an exception occurs, make sure to clean up the skill
+            self.default_shutdown()
+            log_msg = 'Skill initialization failed with {}'
+            LOG.exception(log_msg.format(repr(e)))
```
Comment (chrisveilleux) [inline mycroft/skills/mycroft_skill/mycroft_skill.py:194]: You don't need to include `repr(e)` when using log.exception.

File: mycroft/skills/mycroft_skill/mycroft_skill.py [hunk 4]
```diff
@@ -165,6 +168,32 @@ def __init__(self, name=None, bus=None, use_settings=True):
...
+            raise e
```
Comment (chrisveilleux) [inline mycroft/skills/mycroft_skill/mycroft_skill.py:195]: you don't need to pass the exception to the raise statement

File: mycroft/skills/mycroft_skill/mycroft_skill.py [hunk 5]
```diff
@@ -196,6 +225,20 @@ def _init_settings(self):
...
+        Skill can read the version info from mycroft.version and determine if
+        it can be loaded. This allows the skill v3 framework to disregard
+        incompatible skill classes in a file and select another one.
+
+        Returns:
+            (bool): True if compatible with the version of Mycroft Core,
+                    else False.
+        """
+        return True
+
```
Comment (chrisveilleux) [inline mycroft/skills/mycroft_skill/mycroft_skill.py:241]: I think this needs some discussion.  We are counting on skill developers to keep this up to date with each version?  I think this is a ticking time bomb.  Some skill won't implement it, or implement it incorrectly, and chaos will ensue.

File: mycroft/skills/skill_loader.py [hunk 6]
```diff
@@ -123,11 +124,58 @@ def _get_last_modified_time(path):
         return 0
 
 
+def is_skill_v2(module):
```
Comment (chrisveilleux) [inline mycroft/skills/skill_loader.py:127]: Another versioning thing to track?  Is the "skill framework version" official?  If so, where is it documented?  Is there a way around this?  Will we just keep adding versions upon versions?  When do old versions get cleaned up?

File: mycroft/skills/mycroft_skill/mycroft_skill.py (hunk 2 above)
Comment (forslund) [reply mycroft/skills/mycroft_skill/mycroft_skill.py:171]: This is to handle the current (v2) case which does not pass the bus. The bind, etc will need to be called externally after init.

File: mycroft/skills/mycroft_skill/mycroft_skill.py (hunk 5 above)
Comment (forslund) [reply mycroft/skills/mycroft_skill/mycroft_skill.py:241]: In the general case it should not be overridden by the skill creator, skills should be made compatible with an entire major release and this shouldn't be needed to be implemented.

It is for example for the case if we add a new feature foo in 21.2.7, we want to start using the new feature in the skill. we can then have an `__init__.py`

with

```python
from .skill import SkillWithNewFeature
from .compatibility_skill import CompatibilitySkill
```

compatibility_skill.py

```python
[...]
from mycroft.version import CORE_VERSION_TUPLE

class CompatibilitySkill:
    [...]
    @classmethod
    def is_compatible(cls):
        return CORE_VERSION_TUPLE < (21, 2, 7)
```

And the skill.py:
```python
[...]
from mycroft.version import CORE_VERSION_TUPLE

class SkillWithNewFeature:
```python
from mycroft.version import CORE_VERSION_TUPLE
    [...]
    @classmethod
    def is_compatible(cls):
        return CORE_VERSION_TUPLE >= (21, 2, 7)
```

This has been done earlier with the skill installer skill when the skill-id was added but then the logic was placed in the `create_skill()` function. SInce the suggestion is to remove the need for that method I wanted to suggest a replacement.

The exact same thing can be done with simple imports in the skill __init__

```python
from mycroft.version import CORE_VERSION_TUPLE

if CORE_VERSION_TUPLE < (20, 2, 7):
    from .skill import SkillWithNewFeature
else:
    from .compatibility_skill import CompatibilitySkill
```

My thought was that this was using the classmethod would be more explicit.

File: mycroft/skills/skill_loader.py (hunk 6 above)
Comment (forslund) [reply mycroft/skills/skill_loader.py:127]: The current version is referred to v2 in the skill loader.

We can't drop support for all existing skills without having an alternative in place, so we need to support the current version for some time when making changes. I don't think that more than two versions should ever be supported at the same time. Likely the v2 should be kept for a year or so?

File: mycroft/skills/mycroft_skill/mycroft_skill.py (hunk 4 above)
Comment (forslund) [reply mycroft/skills/mycroft_skill/mycroft_skill.py:195]: Will fix, this part was moved from the old skill loading without any changes

File: mycroft/skills/skill_loader.py (hunk 6 above)
Comment (forslund) [reply mycroft/skills/skill_loader.py:127]: Oh also as touched on above, we can probably skip the compatibility checks if we do some hacking in `__new__()` buuut that's a slippery-slope that I'd rather avoid.

**Comment by JarbasAl** [comment]
> @JarbasAI you're correct this is more of a cleanup. I'll drop the 3 and call it `2_1` in the code next time I update. Regarding the the comments about create_skill couldn't you in most cases just run your "before skill loading" the `super().__init__()` call?
> 
Seems like i missed this back then, my usage was returning a different class based on ovos_utils platform detection, a skill optimized for a voice only experience or GUI experience. 

In particular for the onboarding process (select backend/voice + pair), speaking and asking questions in a loop is annoying but needed in a voice only interface, if we have a GUI we can rely on it and be much less verbose

Other use case would be skills that take advantage of extra features unique to platforms, eg OVOS, chatterbox, bigscreen, neon... or that mock features if missing

Part of this can be options in a single skill, but sometimes it might make sense to have different classes

```python

class VoiceOnlySkill:
    ...
    
class GuiSkill(VoiceOnlySkill):
    ...
    
def create_skill():
     if has_gui():
         return GuiSkill()
     return VoiceOnlySkill()    
     
 ```
 
This can also be done in `__new__` and not used in the wild, so don't take this as a criticism or a blocker, just highlighting a use case, this PR is a good change!

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2855)). 
Mycroft logs are also available: [skills.log](https://reports.mycroft.ai/core/PR-2855/logs/skills.log), [audio.log](https://reports.mycroft.ai/core/PR-2855/logs/audio.log), [voice.log](https://reports.mycroft.ai/core/PR-2855/logs/voice.log), [bus.log](https://reports.mycroft.ai/core/PR-2855/logs/bus.log), [enclosure.log](https://reports.mycroft.ai/core/PR-2855/logs/enclosure.log)

**Comment by forslund** [comment]
There are a couple of ways of handle what you suggest. My thought was the `is_compatible()` class method could be use to define conditions where the skill is/isn't supposed to be used...Might be named differently maybe?

Another possibility could be using imports and if-statements in the `__init__.py`

```python
if has_gui():
    import GuiSkill
else:
    import VoiceOnlySkill
```

File: mycroft/skills/skill_loader.py [hunk 7]
```diff
@@ -150,7 +150,7 @@ def get_skill_class(skill_module):
     """
     for name, obj in skill_module.__dict__.items():
         if (isclass(obj) and issubclass(obj, MycroftSkill) and
-                obj is not MycroftSkill):
+                obj is not MycroftSkill and obj.is_compatible()):
```
Comment (JarbasAl) [inline mycroft/skills/skill_loader.py:153]: what if 2 compatible skills exist in same file? maybe one of them has less features but is still compatible

Comment (forslund) [reply mycroft/skills/skill_loader.py:153]: Right, something like `Alternative Y is also compatible but will not be used`

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2855))

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2855))

**Comment by forslund** [comment]
Closing PR since we're archiving the repo
